provider/rdns.go

```diff
@@ -0,0 +1,549 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/pkg/errors"
+	"istio.io/istio/pkg/log"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+const (
+	rdnsMaxHosts      = 10
+	rdnsOriginalLabel = "originalText"
+	rdnsPrefix        = "/rdnsv3"
+	rdnsTimeout       = 5 * time.Second
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// RDNSClient is an interface to work with Rancher DNS(RDNS) records in etcdv3 backend.
+type RDNSClient interface {
+	Get(key string) ([]RDNSRecord, error)
+	List(rootDomain string) ([]RDNSRecord, error)
+	Set(value RDNSRecord) error
+	Delete(key string) error
+}
+
+// RDNSConfig contains configuration to create a new Rancher DNS(RDNS) provider.
+type RDNSConfig struct {
+	DryRun       bool
+	DomainFilter DomainFilter
+	RootDomain   string
+}
+
+// RDNSProvider is an implementation of Provider for Rancher DNS(RDNS).
+type RDNSProvider struct {
+	client       RDNSClient
+	dryRun       bool
+	domainFilter DomainFilter
+	rootDomain   string
+}
+
+// RDNSRecord represents Rancher DNS(RDNS) etcdv3 record.
+type RDNSRecord struct {
+	AggregationHosts []string `json:"aggregation_hosts,omitempty"`
+	Host             string   `json:"host,omitempty"`
+	Text             string   `json:"text,omitempty"`
+	TTL              uint32   `json:"ttl,omitempty"`
+	Key              string   `json:"-"`
+}
+
+// RDNSRecordType represents Rancher DNS(RDNS) etcdv3 record type.
+type RDNSRecordType struct {
+	Type   string `json:"type,omitempty"`
+	Domain string `json:"domain,omitempty"`
+}
+
+type etcdv3Client struct {
+	client *clientv3.Client
+	ctx    context.Context
+}
+
+var _ RDNSClient = etcdv3Client{}
+
+// NewRDNSProvider initializes a new Rancher DNS(RDNS) based Provider.
+func NewRDNSProvider(config RDNSConfig) (*RDNSProvider, error) {
+	client, err := newEtcdv3Client()
+	if err != nil {
+		return nil, err
+	}
+	domain := os.Getenv("RDNS_ROOT_DOMAIN")
```

Comment:
there are is also types_test.go where you can test ENV Vars

Reply:
Has been added to the [types_test.go](https://github.com/kubernetes-incubator/external-dns/pull/1098/files#diff-e964ba9ac8ef2ae87c34772f0459e71fR343) file.